WIP Add basic architecture docs

lucyleeow opened the pull request:
# Description
This is a still a WIP and is a PR to my own fork just to get some feedback.

So far have added docs about how our python model classes/event emission works using the great unmerged PR: https://github.com/napari/napari/pull/1965/ and info from: https://github.com/napari/napari/issues/1436

To do:
* add more detail about qt classes - thinking about how best to do this without overlap with docs: [event loop docs](https://napari.org/guides/event_loop.html) and [hooking up your own events](https://napari.org/howtos/connecting_events.html)
* add more detail about vispy
* add diagram about how vispy, qt and python connect

## Type of change
<!-- Please delete options that are not relevant. -->
- [ ] Bug-fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# References
<!-- What resources, documentation, and guides were used in the creation of this PR? -->
<!-- If this is a bug-fix or otherwise resolves an issue, reference it here with "closes #(issue)" -->

# How has this been tested?
<!-- Please describe the tests that you ran to verify your changes. -->
- [ ] example: the test suite for my feature covers cases x, y, and z
- [ ] example: all tests pass with my change
- [ ] example: I check if my changes works with both PySide and PyQt backends
      as there are small differences between the two Qt bindings.  

## Final checklist:
- [ ] My PR is the minimum possible work for the desired functionality
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] If I included new strings, I have used `trans.` to make them localizable.
      For more information see our [translations guide](https://napari.org/developers/translations.html).

lucyleeow committed:
add arch docs

lucyleeow committed:
add title;

lucyleeow committed:
typo

lucyleeow committed:
use current step

lucyleeow committed:
use model examples

dependabot[bot] committed:
ci(dependabot): bump docker/login-action from 1.9.0 to 2 (#4774)

Bumps [docker/login-action](https://github.com/docker/login-action) from 1.9.0 to 2.
- [Release notes](https://github.com/docker/login-action/releases)
- [Commits](https://github.com/docker/login-action/compare/28218f9b04b4f3f62068d7b6ce6ca5b26e35336c...49ed152c8eca782a232dede0303416e8f356c37b)

---
updated-dependencies:
- dependency-name: docker/login-action
  dependency-type: direct:production
  update-type: version-update:semver-major
...

Signed-off-by: dependabot[bot] <support@github.com>

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
ci(dependabot): bump bruceadams/get-release from 1.2.2 to 1.2.3 (#4773)

Bumps [bruceadams/get-release](https://github.com/bruceadams/get-release) from 1.2.2 to 1.2.3.
- [Release notes](https://github.com/bruceadams/get-release/releases)
- [Commits](https://github.com/bruceadams/get-release/compare/v1.2.2...v1.2.3)

---
updated-dependencies:
- dependency-name: bruceadams/get-release
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
ci(dependabot): bump actions/github-script from 5 to 6 (#4772)

Bumps [actions/github-script](https://github.com/actions/github-script) from 5 to 6.
- [Release notes](https://github.com/actions/github-script/releases)
- [Commits](https://github.com/actions/github-script/compare/v5...v6)

---
updated-dependencies:
- dependency-name: actions/github-script
  dependency-type: direct:production
  update-type: version-update:semver-major
...

Signed-off-by: dependabot[bot] <support@github.com>

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
ci(dependabot): bump actions/download-artifact from 2 to 3 (#4771)

Bumps [actions/download-artifact](https://github.com/actions/download-artifact) from 2 to 3.
- [Release notes](https://github.com/actions/download-artifact/releases)
- [Commits](https://github.com/actions/download-artifact/compare/v2...v3)

---
updated-dependencies:
- dependency-name: actions/download-artifact
  dependency-type: direct:production
  update-type: version-update:semver-major
...

Signed-off-by: dependabot[bot] <support@github.com>

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
ci(dependabot): bump styfle/cancel-workflow-action from 0.9.1 to 0.10.0 (#4770)

Bumps [styfle/cancel-workflow-action](https://github.com/styfle/cancel-workflow-action) from 0.9.1 to 0.10.0.
- [Release notes](https://github.com/styfle/cancel-workflow-action/releases)
- [Commits](https://github.com/styfle/cancel-workflow-action/compare/0.9.1...0.10.0)

---
updated-dependencies:
- dependency-name: styfle/cancel-workflow-action
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

ctrueden committed:
Make references to examples link there (#4767)

* docs/index: fix references to main branch

* Make references to examples folder link there

brisvag committed:
fix clim init for dask arrays (#4724)

* fix clim init for dask arrays

* fix tensorstore

* avid double-calculation

* add test for clim

* move test to dask and make sync only

* use normalize_dtype

DragaDoncila committed:
Set focus on `QtViewer` object after creating main window (#4768)

Czaki committed:
Add missed call of `running_as_bundled_app` in `__main__` (#4777)

cnstt committed:
Add public API for dims transpose (#4727)

* Update dims.py

Add public API for dims transpose

* [pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

* Remove _transpose usage, deprecate it and fix typos

* Fix typo

* Update napari/components/dims.py

Co-authored-by: Kevin Yamauchi <kevin.yamauchi@gmail.com>

* [pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

* Update dims.py

Remove old _transpose private method

Co-authored-by: pre-commit-ci[bot] <66853113+pre-commit-ci[bot]@users.noreply.github.com>
Co-authored-by: Matthias Bussonnier <bussonniermatthias@gmail.com>
Co-authored-by: Kevin Yamauchi <kevin.yamauchi@gmail.com>

tlambert03 committed:
Use `npe2 list` to show plugin info (#4739)

* use npe2 list for plugin info

* bump npe2

* fix test

* fix test

Czaki committed:
Fix source of copied layer events by use `layer.as_layer_data_tuple` (#4681)

JoOkuma committed:
Fixing readme typo (#4791)

tlambert03 committed:
test: fix ability to use posargs with tox (#4788)

* test: fix tox posargs

* change example skip approach

* add explicit request

Czaki committed:
Refactor `NapariQtNotification` test for better cleaning Qt objects (#4763)

andy-sweet committed:
Use custom color field classes in all models (#4704)

* Move ColorValue and ColorArray into new module

* Use color field types in color manager

* Use color field in categorical colormap

* Use color field in wireframe and normals

* Add docs to custom field types.

* Fix docstring

* Use validate instead of validate_type

* Revert changes to color manager

* Remove TODOs

* Update docstring with examples

* Use ColorValue in scale bar

* Fix ColorArray validator

* Clean up docstrings.

melissawm committed:
Fix documentation links errors. (#4787)

jni committed:
Add Talley to the steering council (#4798)

lucyleeow committed:
add qt vispy sections

lucyleeow commented:
Added minimal sections on qt and vispy (not sure what else should/could be added)
Am thinking of adding a diagram about how python / qt / vispy models are connected.

andy-sweet commented on docs/developers/napari_architecture.md:
```diff
@@ -0,0 +1,164 @@
+Commonly, python models in napari are classes that store information about their
+state as an attribute and are the "source of ground truth". When these
+attributes are changed an "event" needs to be emitted such that relevant
+obserers of the model (such as other classes) can take the appropriate
```
```suggestion
observers of the model (such as other classes) can take the appropriate
```

andy-sweet commented on docs/developers/napari_architecture.md:
```diff
@@ -0,0 +1,164 @@
+state as an attribute and are the "source of ground truth". When these
+attributes are changed an "event" needs to be emitted such that relevant
+obserers of the model (such as other classes) can take the appropriate
+action. See [An Introduction to the Event Loop in napari](events/event_loop))
```
I would probably avoid linking to the event loop to avoid confusion between the python model's events and the napari/Qt event loop.

I think you describe the python model events well as what is effectively a simple implementation of [the observer pattern](https://en.wikipedia.org/wiki/Observer_pattern).

One critical difference between the python model events and the Qt event loop is that (by default) the model events will execute connected callbacks immediately on the same thread that the event is emitted.

Qt events (e.g `MouseEvent`) will be queued up to be processed on their event loop (in napari, likely the event loop) on the main thread. Similarly signals/slots connected to signals may be also be queued in this way depending on the connection type.

It might be worth linking out to [the events reference page though](https://napari.org/guides/events_reference.html), which has a small blurb about what napari events are.

lucyleeow replied:
Thanks for that, I never knew. Can you expand - when is the python model event used and when is the Qt event loop used?

andy-sweet replied:
To clarify, when you're running the napari app (i.e. not headless), then pretty much everything is running the Qt main event loop, except when you explicitly run things in another thread (which isn't often right now). That's what the docs you original pointed to start to explain.

[QEvent](https://doc.qt.io/qt-5/qevent.html)s will mostly come from the underlying OS/windowing system to indicate things like mouse moves/clicks, keyboard presses, hiding/showing. The standard way to handle these events is to implement/override `QWidget` event handling methods (e.g. [QWidget::mouseMoveEvent](https://doc.qt.io/qt-5/qwidget.html#mouseMoveEvent)). The emission of these events is not something that napari code should explicitly do, though calling some methods may trigger them.

napari's events (defined by `Event`, `EventEmitter`, and `EmitterGroup`) are independent of `QEvents` and don't need an event loop to be emitted or processed. For example, `Image.events` is an `EmitterGroup`, and calling `Image.events.set_data` (which is an `EventEmitter`) will just immediately and synchronously execute any callbacks that were attached to it using `connect`. These are used explicitly in napari, and are often public, meaning that plugins and other things that depend on napari may connect to them.

Qt signals/slots are another topic to not get confused by here. The `connect` method is also crucial, but they have their own Qt and Qt event-loop specific logic that is better explained by the Qt docs.

lucyleeow replied:
Sorry, some follow up questions (just so I understand):
> The standard way to handle these events is to implement/override QWidget event handling methods 

Does this mean that QEvents are overridden by QWidget event handling methods?

> napari's events (defined by Event, EventEmitter, and EmitterGroup) are independent of QEvents and don't need an event loop to be emitted or processed.

and 

> when you're running the napari app (i.e. not headless), then pretty much everything is running the Qt main event loop

I don't follow this - napari events are independent of QEvents, but they will still be run via the Qt main event loop when running the napari app?

andy-sweet replied:
> I don't follow this - napari events are independent of QEvents, but they will still be run via the Qt main event loop when running the napari app?

In a Qt app, everything is run on the main event loop (once it has started) unless explicitly run on a separate thread. Another way to think about this is that the Qt event loop is the only thing that runs on the main thread of napari once it has started. It is either processing `QEvents` (which could trigger other execution like the emission of a napari event) or it is idling and waiting for the next `QEvent`.

That's my understanding anyway, but it's consistent with [the small blurb on napari.org](https://napari.org/stable/guides/event_loop.html#more-in-depth).

andy-sweet commented on docs/developers/napari_architecture.md:
```diff
@@ -0,0 +1,164 @@
+not the layer models although there is intention to convert these to
+`EventedModels` in the future.
+
+## Qt models
```
I'd avoid calling these Qt *models* - maybe just Qt classes instead? That helps to avoid confusion with napari's app model (since most Qt classes are not part of that), Qt model classes (e.g. [QAbstractItemModel](https://doc.qt.io/qt-6/qabstractitemmodel.html)) some of which we do actually use, though as an implementation detail.

andy-sweet commented on docs/developers/napari_architecture.md:
```diff
@@ -0,0 +1,164 @@
+Qt classes are responsible for all napari's user interface elements. There is
+generally one to one mapping between Python models and qt models in napari, for
+example Python model `Dims` and Qt model `QtDims`.
+The Qt class can register callbacks such that when an attribute of the
```
If I understand correctly, most of the Qt classes (at least those that implement GUI components), are intended to be the *view* (and often controller) part of the application model, which is roughly model-view-controller.

You effectively mention this at the top of the document, so explicitly tying this section back to that summary seems like a good idea.

andy-sweet replied:
Adding an example of `the appropriate actions` here would be great. E.g. `dims.current_step = (5, 0, 0)` will cause the `QtDims` object to update the position of the slider representing the first dimesion.

lucyleeow replied:
To clarify, does 'application model' mean 'model viewer and controller' ?

andy-sweet replied:
Oops, I shouldn't have said application *model* - that's just confusing - but yes you're right. Maybe *application architecture* would be better as MVC is an architectural pattern.

andy-sweet commented on docs/developers/napari_architecture.md:
```diff
@@ -0,0 +1,164 @@
+the Python model, which gets updated directly when a field is changed via the
+GUI.
+
+## Vispy models
```
Again, I'd avoid using the world *models* for similar reasons as the Qt section. I think classes or *Vispy layers* would be fine.

If I understand napari's architecture correctly, the vispy classes represent a view on the layers (which represent a part of the core model/data of the app) - like you mentioned that specifically updates the canvas (which in the case of napari right now does have a Qt widget backing it). In some sense the vispy classes also represent a model for data that will be sent to the GPU, but that's probably a bit nuanced to include here.

andy-sweet commented on docs/developers/napari_architecture.md:
```diff
@@ -0,0 +1,164 @@
+# Napari architecture
```
I really like this opening section. It sets up the rest of the document well. Maybe a small/simple diagram that depicts some of the organization described below would be useful?

andy-sweet commented on docs/developers/napari_architecture.md:
````diff
@@ -0,0 +1,164 @@
+Using `EventedModel` would reduce the above `weather` class code to:
+
+```python
+class weather(EventedModel):
````
```suggestion
class Weather(EventedModel):
```

andy-sweet commented on docs/developers/napari_architecture.md:
````diff
@@ -0,0 +1,164 @@
+```python
+from napari.utils.events import EmitterGroup
+
+class Weather:
````
This example does a great job of explaining the Python models in napari and what boilerplate `EventedModel` absorbs! But it seems harder to reuse in the qt/vispy sections, which would be ideal. Maybe I'm being a boring pragmatist, but I think using a simplified version of one of the existing napari models could be a good idea. I think `Dims` might be a good target (but that's probably because I have dims/slicing on the mind) - that way you could show the slider in the Qt section and the canvas update in the vispy section.

Other alternatives could be `ViewerModel` and `LayerList`, though those might be a bit more complex.

lucyleeow replied:
I had something like that but changed it to this. I like how the example shows how the same thing could be done with getter/setter and EventedModel base class. I guess I could take a Dims example but first show how it would be done with getter/setters.

andy-sweet replied:
Yeah, it should be pretty easy to translate from the `Dims` model to the property getter/setter (I think that's what it used to be). It would probably be a good idea to use a simplified version of `Dims` (e.g. maybe with only `ndim`, `ndisplay`, and `current_step`). Sometimes I put `...` in doc code examples to indicate there is more in a class/function, but I'm not going to talk about it here because it's not relevant.

For Qt, we could talk about `QtDims` and `QtDimSliderWidget` event, commenting on how that connects to `current_step`. And for vispy, the effect of changing `ndisplay` might be a good one.

andy-sweet reviewed (commented):
Looks good so far - hope the comments are useful.

tlambert03 committed:
mock npe1 plugin manager too during tests (#4806)

jaimergp committed:
Discuss the approval of NAP-2 (conda-based packaging for napari) (#4602)

Co-authored-by: Talley Lambert <talley.lambert@gmail.com>
Co-authored-by: Juan Nunez-Iglesias <juan.nunez-iglesias@monash.edu>

mstabrin committed:
Allow stack to be specified multiple times (#4783)

Co-authored-by: Markus Stabrin <markus.stabrin@gmail.com>

github-actions[bot] committed:
[Automatic] Update albertosottile/darkdetect vendored module (#4808)

Co-authored-by: Carreau <Carreau@users.noreply.github.com>
Co-authored-by: Juan Nunez-Iglesias <juan.nunez-iglesias@monash.edu>

jni committed:
Accept NAP-2: Distributing napari with conda-based packaging (#4810)

mstabrin committed:
Improve "select all" logic in points layer, adding "select all across slices" behavior (#4809)

melissawm committed:
Fixes duplicate label warnings in the documentation (#4823)

jamesyan-git committed:
Make newly added points editable (#4829)

* newly added points now editable

* added tests

* parameterized tests

* added documentation

* [pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

* tidy up

* edge_width test added, points name removed, typo fixed

* [pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

Co-authored-by: pre-commit-ci[bot] <66853113+pre-commit-ci[bot]@users.noreply.github.com>

Commit:
add update_mesh method (#4793)

* add update_mesh method

* cleanup additional usage, add displayed_stored to update_mesh

* apply reviewer suggestions

Co-authored-by: Lorenzo Gaifas <brisvag@gmail.com>

Co-authored-by: Kim Pevey <kpevey@quansight.com>
Co-authored-by: Lorenzo Gaifas <brisvag@gmail.com>

Czaki committed:
Improve using Qt flags by direct use enums. (#4817)

cajongonzales committed:
Allow empty tuple layer to be returned #4571 (#4831)

* Allow empty tuple layer to be returned #4571

An empty list can sometimes be returned from adding a layer.

* Add test for misc ensure_list_of_layer_data_tuple

* [pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

Co-authored-by: pre-commit-ci[bot] <66853113+pre-commit-ci[bot]@users.noreply.github.com>

Czaki committed:
Embeding viewers in plugins widget example (#4665)

Co-authored-by: cnstt <104427648+cnstt@users.noreply.github.com>
Co-authored-by: Andy Sweet <andrew.d.sweet@gmail.com>
Co-authored-by: Lorenzo Gaifas <brisvag@gmail.com>

Czaki committed:
dissable notifications if no main window (#4839)

tlambert03 committed:
support PyQt6, remove logic for `compile_qt_svgs` (#3707)

* no longer compile resources

* more fixes

* misc other updates

* use qtpy main

* block signal when changing

* undo qtpy change

* remove print

* update qss

* more styles fixes

* update gitignore

* fix overload

* fix image reshape for pyside6

* fix more tests in _qt

* fix pos()

* fix examples tests

* remove icons altogether

* fix headless

* fix menus and globalPosition

* fix pyside6

* fix last pyqt6 test

* Update napari/__main__.py

Co-authored-by: Grzegorz Bokota <bokota+github@gmail.com>

* unskip console tests on qt6

* Update napari/_qt/qt_main_window.py

Co-authored-by: Andy Sweet <andrew.d.sweet@gmail.com>

* Update napari/resources/_icons.py

Co-authored-by: Andy Sweet <andrew.d.sweet@gmail.com>

* comments and fixes

* go back to bool

* add comment

* add back fix for point in scrollbar

Co-authored-by: Grzegorz Bokota <bokota+github@gmail.com>
Co-authored-by: Andy Sweet <andrew.d.sweet@gmail.com>

jaimergp committed:
add 'napari-xpra' target to dockerfile (#4703)

* add 'napari-xpra' target to dockerfile

* add instructions for developers

* temporarily add pull_request trigger

* weird typo

* skimage not needed

* Do not run on PRs [ci skip]

melissawm committed:
Add versioned API docs workflow (#4635)

Czaki committed:
small cleanup for NapariQtNotification code (#4841)

brisvag committed:
Allow dunder methods use in `PublicOnlyProxy` (#4792)

tlambert03 committed:
Raise better errors from lazy __getattr__ (#4848)

gselzer committed:
Add the ability to return a `List[Layer]` in magicgui (#4851)

* Add the ability to return a List[Layer]

* Add documentation on List[Layer] returns

lucyleeow committed:
review comments

lucyleeow commented on docs/developers/napari_architecture.md:
```diff
@@ -0,0 +1,198 @@
+    """Vipsy camera for both 2D and 3D rendering.
+    """
+
+    def __init__(self):
```
@andy-sweet silly question - do classes need to instantiate with `dims` attribute to be able to connect a callback?

lucyleeow committed:
amend vispycamera

Czaki committed:
Update base docker image to ubuntu 22.04, optimize container size. (#4568)

goanpeca committed:
Fix deepcopy for trans objects when passing args (#4786)

* Fix deepcopy for trans objects when passing args

* Unify trans deepcopy tests

* [pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

* Update napari/utils/_tests/test_translations.py

Co-authored-by: Andy Sweet <andrew.d.sweet@gmail.com>

Co-authored-by: pre-commit-ci[bot] <66853113+pre-commit-ci[bot]@users.noreply.github.com>
Co-authored-by: Andy Sweet <andrew.d.sweet@gmail.com>

github-actions[bot] committed:
Update vendored modules. (#4845)

Co-authored-by: Carreau <Carreau@users.noreply.github.com>

cnstt committed:
Update documenation : install plugin from URL  (#4799)

* Update find_and_install_plugin.md

Update plugin install via url documentation

* Update find_and_install_plugin.md

* rephrase, move admonition

Co-authored-by: Talley Lambert <talley.lambert@gmail.com>

Carreau committed:
Add mini interactive prompt to edit string_list.json (#4733)

This adds a mini TUI that ask for each non translated string whether to
ignore it or not, allowing to quickly update string_list.json.

This implements a quick getch() function that read stdin for a single
character, but will only work on unix.

This roughly looks like this in a loop.

```
    i : ignore –  add to ignored localised strings
    q : quit –  quit w/o saving
    s : save and quit

    napari/_qt/dialogs/qt_plugin_dialog.py:700 'unavailable'

                      )
                  )
    >             widget.setObjectName("unavailable")
                  widget.style().unpolish(widget)
                  widget.style().polish(widget)
                  widget.action_button.setEnabled(False)
```

Alternate/addition to #4732

Carreau committed:
Cleanup list of non-translatable strings. (#4732)

All those strings seem to not exists in the corresponding files.

Carreau committed:
Update deprecation warnings with versions (#4757)

* Update deprecation warnings with versions

Subset of #4756 that should be non-controversial.

* Apply suggestions from code review

* Update napari/layers/image/image.py

Co-authored-by: Andy Sweet <andrew.d.sweet@gmail.com>

Co-authored-by: Andy Sweet <andrew.d.sweet@gmail.com>

brisvag committed:
Bump vispy 0.11.0 (#4778)

cnstt committed:
Update hub.py (#4863)

* Update hub.py

* [pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

* Update hub.py

* Update hub.py

* Update hub.py

* [pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

Co-authored-by: pre-commit-ci[bot] <66853113+pre-commit-ci[bot]@users.noreply.github.com>

kne42 committed:
add path specificity scoring (#4830)

Score an fnmatch pattern, with higher specificities having lower scores.

Absolute paths have highest specificity, followed by paths with the most nesting, then by path segments with the least ambiguity.

andy-sweet committed:
Fix multiscale level selection when zoomed out (#4842)

melissawm committed:
Fix NAP-3 table of contents (#4872)

Czaki committed:
Restore Labels, Points and Shapes layer mode after hold key (#4814)

Co-authored-by: Lorenzo Gaifas <brisvag@gmail.com>

lucyleeow committed:
review changes

lucyleeow commented:
Very simplistic idea of what a diagram could look like, happy to take suggestions. I could take a specific example of e.g., a `Dims` attribute being changed via GUI and how this would affect the other classes but this felt more general/neater

![napari](https://user-images.githubusercontent.com/23182829/181418092-4221c301-5fd6-49b6-8d86-2dd913bc36d2.png)

rwkozar committed:
added description of how to save layers without compression (#4832)

* added description of how to save layers with compression

* added instructions for saving with non-default compression

* updated links to documentation

* added message about saving without compression to image howto

* Update docs/howtos/layers/image.md

Co-authored-by: Draga Doncila Pop <17995243+DragaDoncila@users.noreply.github.com>

Carreau committed:
MAINT: disable failing conda bundling. (#4878)

Until fixed, otherwise all PRs everywhere are failing.
Please open an issue to track re-enabling when merging this PR.

Czaki committed:
Fix creation `QtViewer` for `ViewerModel` with already added layers (#4854)

* fix creation viewer for empty model

* attempt to clean gc references

Co-authored-by: Matthias Bussonnier <bussonniermatthias@gmail.com>

Carreau committed:
MAINT: Work around conda bundling issue. (#4883)

Follow-up #4878, May need update to #4879

Czaki committed:
Fix viewer.reset() throws TypeError (#4868)

Fix raising error in napari Viewer on call reset method.

Czaki committed:
Confirmation on close, alternative approach (#4820)

* confirm on close vol2

* add confirmation on close by alt+f4 or close button

* add tests

* fix tests

tlambert03 committed:
fix icon urls (#4887)

alisterburt committed:
Expose points layer antialiasing control publicly (#4735)

Co-authored-by: Grzegorz Bokota <bokota+github@gmail.com>
Co-authored-by: Kevin Yamauchi <kevin.yamauchi@gmail.com>
Co-authored-by: Lorenzo Gaifas <brisvag@gmail.com>

tlambert03 committed:
Refactor: use app-model and in-n-out (#4784)

* feat: use app-model WIP

* partially fixed circ imports

* bring back some menus with lazy import

* wip

* incorporate changes from app-model and ino

* remove old code

* remove more

* add app-model dep

* remove module from typecheck

* small changes

* remove not gc.collect assertion

* reduce test code

* add comments about context objects

* add in-n-out to deps

* [pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

* fix test

* bump in-n-out min ver

* refactor: add constants module, add test for menu building

* add __all__ to constants

* add some tests

* add more comments

* bump app-model

* add internal documentation

* add motivations and vision

* bump version, raise synchronous errors

* comment navigation global

* lenient test

* Update docs/guides/app_model.md

Co-authored-by: alisterburt <alisterburt@gmail.com>

* Update napari/_app/actions/_layer_actions.py

Co-authored-by: alisterburt <alisterburt@gmail.com>

* Update napari/_qt/widgets/_tests/test_qt_extension2reader.py

Co-authored-by: Grzegorz Bokota <bokota+github@gmail.com>

* rename module

* use get_app getter

* bump appmodel

* fix comment

* add comment

* move layer actions constants

* clarify expression

* update docs

* add tests

* update docs

* add comment to layer actions

* add comment to named tuple

* more comments for nathan

* add comment

* remove in-n-out, pin app-model

* Update docs/guides/app_model.md

Co-authored-by: Kira Evans <contact@kne42.me>

* Update docs/guides/app_model.md

Co-authored-by: Kira Evans <contact@kne42.me>

Co-authored-by: pre-commit-ci[bot] <66853113+pre-commit-ci[bot]@users.noreply.github.com>
Co-authored-by: alisterburt <alisterburt@gmail.com>
Co-authored-by: Grzegorz Bokota <bokota+github@gmail.com>
Co-authored-by: Kira Evans <contact@kne42.me>

dependabot[bot] committed:
ci(dependabot): bump toshimaru/auto-author-assign from 1.5.0 to 1.6.1 (#4890)

Bumps [toshimaru/auto-author-assign](https://github.com/toshimaru/auto-author-assign) from 1.5.0 to 1.6.1.
- [Release notes](https://github.com/toshimaru/auto-author-assign/releases)
- [Changelog](https://github.com/toshimaru/auto-author-assign/blob/main/CHANGELOG.md)
- [Commits](https://github.com/toshimaru/auto-author-assign/compare/v1.5.0...v1.6.1)

---
updated-dependencies:
- dependency-name: toshimaru/auto-author-assign
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

brisvag committed:
Fix missing slice from dims (#4889)

tlambert03 committed:
Split out builtins into another top-level module (#4706)

* wip - splitting out builtins

* working read tests

* small changes to reader tests

* wip

* builtin tests working

* [pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

* remove need for pooch

* more reduction

* fix precommit

* fix minreqs

* more coverage

* fix manifest

* minreqs

* add builtins to pip test

* remove napari from test

* return one test to napari

* don't use LayerList

* remove more napari

* update test

* fix conftest

* try split pyargs tests

Co-authored-by: pre-commit-ci[bot] <66853113+pre-commit-ci[bot]@users.noreply.github.com>

tlambert03 committed:
remove a few more qtbot.wait calls in tests (#4888)

* remove a bunch of qtbot.wait

* undo changes in plugin dialog

kcpevey committed:
Optimization: convert vector slice indices to floats (#4794)

* define dtype

* start sorting out how to combine with points by commenting current code

* adding comments for understanding

* final cleanup

* Update napari/layers/vectors/vectors.py

Co-authored-by: Andy Sweet <andrew.d.sweet@gmail.com>

* move code into if statement

Co-authored-by: Kim Pevey <kpevey@quansight.com>
Co-authored-by: Andy Sweet <andrew.d.sweet@gmail.com>

andy-sweet committed:
NAP 4: asynchronous slicing (#4892)

* Initial commit

* Clean up

* Start detailed description

* Add some notes on shapes layer state. (#17)

* Add some notes on shapes layer state.

* Update docs/naps/4-async-slicing.md

Co-authored-by: Andy Sweet <andrew.d.sweet@gmail.com>

* Update docs/naps/4-async-slicing.md

Co-authored-by: Andy Sweet <andrew.d.sweet@gmail.com>

* Update docs/naps/4-async-slicing.md

Co-authored-by: Andy Sweet <andrew.d.sweet@gmail.com>

* Update docs/naps/4-async-slicing.md

Co-authored-by: Andy Sweet <andrew.d.sweet@gmail.com>

* Update docs/naps/4-async-slicing.md

Co-authored-by: Andy Sweet <andrew.d.sweet@gmail.com>

* Cleanup of notes on Shapes layer.

Co-authored-by: Andy Sweet <andrew.d.sweet@gmail.com>

* Vector async slicing nap (#18)

* add_vector_info

* Apply suggestions from code review

Co-authored-by: Andy Sweet <andrew.d.sweet@gmail.com>

* add author

Co-authored-by: Kim Pevey <kpevey@quansight.com>
Co-authored-by: Andy Sweet <andrew.d.sweet@gmail.com>

* Some simplification

* Update NAP

* Revert "Update NAP"

This reverts commit e92077c18dc0be1c7742315fb263dce0e95f26f7.

* Update existing and proposed designs

* Add future work and alternatives

* Clean up alternatives

* Clean formatting.

* Pre-PR cleanup

* More clean up

* Update docs/naps/4-async-slicing.md

Co-authored-by: Nicholas Sofroniew <sofroniewn@gmail.com>

* Address PR feedback

* Update abstract

* Clean up abstract

* Update motivating video

* Signal to event

* Add discussion for initial PR

Co-authored-by: Eric Perlman <perlman@gmail.com>
Co-authored-by: Kim Pevey <kcpevey@quansight.com>
Co-authored-by: Kim Pevey <kpevey@quansight.com>
Co-authored-by: Nicholas Sofroniew <sofroniewn@gmail.com>

Czaki committed:
Make `viewer.events.layers_change` event deprecated (#4895)

* make event deprecated

* fix emmiting event

jaimergp committed:
Docker: Update xpra's apt too (#4901)

* Docker: Update xpra's apt too

* enable PR trigger temporarily

* undo PR triggers [skip ci]

kephale committed:
Raise exception when layer created with a scale value of 0 (#4906)

* Raise exception when layer created with a scale value of 0

* When checking for 0 values in scale also check if scale is None

* Update zero scale check to handle numpy/dask/xarray types

* Move test to napari/layers/_tests; change exception to ValueError

* Clean quoting in exception string

* Change test to use array with zeros instead of random values

kandarpksk committed:
Correct minor typo in quick_start.md (#4908)

codemonkey800 committed:
Add Google calendar ID to directive (#4914)

Commit:
Minor type fix (add Optional) (#4916)

melissawm commented on docs/developers/napari_architecture.md:
```diff
@@ -0,0 +1,200 @@
+* qt classes that handle the interactive GUI aspect of the napari viewer
+    * the private qt code lives in `napari/_qt` and the smaller public qt
+      interface code lives in `napari/qt`
+* vispy classes that handles rendering
```
```suggestion
* vispy classes that handle rendering
```

melissawm commented on docs/developers/napari_architecture.md:
```diff
@@ -0,0 +1,200 @@
+Commonly, python models in napari are classes that store information about their
+state as an attribute and are the "source of ground truth". When these
+attributes are changed an "event" needs to be emitted such that relevant
+obsevers of the model (such as other classes) can take the appropriate
```
```suggestion
observers of the model (such as other classes) can take the appropriate
```

melissawm commented on docs/developers/napari_architecture.md:
```diff
@@ -0,0 +1,200 @@
+* type validation and coersion on class instantiation and attribute assignment
+* event emission after successful attribute assignment
+
+Using `EventedModel` would reduce the above `Dim` class code to:
```
In the previous example, the callback made it clear what was happening when something is changed. Could you clarify how this works with `EventedModel`? Do you still need to connect the callback with this new Dim definition?

lucyleeow replied:
Ah yes good point!

melissawm reviewed (commented):
Hi @lucyleeow - this is really helpful!

lucyleeow committed:
address review

lucyleeow commented:
I now realise 'napari architecture' may not be the best name since this doc only talks about 'python models and events' - maybe this would be a better title?

melissawm commented:
I'd go for something like "napari models and events", since you mention 3 categories here - python, qt and vispy.

jaimergp committed:
Conda: fix MacOS signing (#4904)

* Conda: remove clobbering 'codesign' on osx

* temporarily enable PR triggers for signing

* disable PR triggers for signing

pre-commit-ci[bot] committed:
[pre-commit.ci] pre-commit autoupdate (#4779)

Co-authored-by: pre-commit-ci[bot] <66853113+pre-commit-ci[bot]@users.noreply.github.com>
Co-authored-by: Grzegorz Bokota <bokota+github@gmail.com>

lucyleeow committed:
rename

tlambert03 committed:
use npe2api in plugin install dialog (#4893)

* use npe2api by default

* add typing

* fix import

* [pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

* remove old test

* fix typing again

* add user agent

Co-authored-by: pre-commit-ci[bot] <66853113+pre-commit-ci[bot]@users.noreply.github.com>

tlambert03 committed:
Enable plugin menu contributions with app-model (#4847)

* feat: use app-model WIP

* partially fixed circ imports

* bring back some menus with lazy import

* wip

* incorporate changes from app-model and ino

* remove old code

* remove more

* add app-model dep

* remove module from typecheck

* small changes

* remove not gc.collect assertion

* reduce test code

* add comments about context objects

* add in-n-out to deps

* [pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

* fix test

* bump in-n-out min ver

* refactor: add constants module, add test for menu building

* add __all__ to constants

* add some tests

* add more comments

* bump app-model

* add internal documentation

* add motivations and vision

* bump version, raise synchronous errors

* wip

* wip: working on evented menu contribs

* move some stuff

* minimize

* move plugin init

* comment navigation global

* lenient test

* Update docs/guides/app_model.md

Co-authored-by: alisterburt <alisterburt@gmail.com>

* Update napari/_app/actions/_layer_actions.py

Co-authored-by: alisterburt <alisterburt@gmail.com>

* Update napari/_qt/widgets/_tests/test_qt_extension2reader.py

Co-authored-by: Grzegorz Bokota <bokota+github@gmail.com>

* rename module

* use get_app getter

* bump appmodel

* fix comment

* add comment

* move layer actions constants

* clarify expression

* update docs

* move is_menu_contributable

* add tests

* bump npe2 version

* fix test

* add more tests

* fix for py38

* update docs

* add comment to layer actions

* add comment to named tuple

* more comments for nathan

* add comment

* fix mypy

* remove in-n-out, pin app-model

* Update docs/guides/app_model.md

Co-authored-by: Kira Evans <contact@kne42.me>

* Update docs/guides/app_model.md

Co-authored-by: Kira Evans <contact@kne42.me>

* add comment

* fix test

Co-authored-by: pre-commit-ci[bot] <66853113+pre-commit-ci[bot]@users.noreply.github.com>
Co-authored-by: alisterburt <alisterburt@gmail.com>
Co-authored-by: Grzegorz Bokota <bokota+github@gmail.com>
Co-authored-by: Kira Evans <contact@kne42.me>

DragaDoncila commented on docs/developers/napari_models.md:
```diff
@@ -0,0 +1,226 @@
+The napari codebase can be thought to consist of three main components:
+
+* python models describing objects - these are able to operate separately from
+  the viewer and do not have any dependencies on user interface classes
```
Hmm we should be careful here about what we refer to as "the viewer" since there is a `ViewerModel`. Maybe we say "able to operate without the GUI interface" or something similar. I would expect this documentation to serve current/prospective developers so I think using GUI is fine.

DragaDoncila commented on docs/developers/napari_models.md:
```diff
@@ -0,0 +1,226 @@
+
+The separation of the python models from viewer GUI code allows:
+
+* the python model to be easily run headless without the viewer, for example
```
Might be good to expand on the definition of 'headless' in a little info admonition and maybe provide a small code example that illustrates how you could use napari in headless mode.

lucyleeow replied:
I realise don't actually know how to use napari in headless mode.... :grimacing:

lucyleeow replied:
I couldn't find anything in the docs, maybe it would also be a good addition to the documentation (maybe a page in the 'getting started' section..)

tlambert03 replied:
i'd say it's more of a "concept" at this point :joy: ... with probably not a huge amount of use cases.  But it's something that we generally protect (in terms of keeping the gui isolated) out of hopes for more future usage.

Headless mode is pretty much just anything without showing the viewer
```python
viewer = napari.Viewer(show=False)

viewer_model = napari.components.ViewerModel()
```

but one could also directly use lower level components like the `LayerList` and `Layer` objects (without a viewer).

I'll note that currently, using `Viewer(show=False)` still won't prevent qt stuff from being imported... however, if one uninstalls qtpy or any of the qt components, then it will work just fine (and qt won't be initialized).

this is all something that's good to know for a napari dev, but I'm not sure it's something that someone getting started with napari would need to understand (or should be exposed to just yet).

DragaDoncila replied:
Yeah agreed not necessarily needed for a user but since this is developer facings docs, I'd say it's worth even just summarising @tlambert03's comment above in a little admonition box

lucyleeow replied:
Thanks! I will update the docs with the info

DragaDoncila commented on docs/developers/napari_models.md:
````diff
@@ -0,0 +1,226 @@
+        self.events.ndisplay(value=value)
+```
+
+Another object can then "listen" for changes in our `Dim` model and register
````
```suggestion
Another object can then "listen" for changes in our `Dims` model and register
```

DragaDoncila commented on docs/developers/napari_models.md:
````diff
@@ -0,0 +1,226 @@
+
+```python
+# create an instance of the model
+dims = Dim(ndim=3, ndisplay=2)
````
```suggestion
dims = Dims(ndim=3, ndisplay=2)
```

DragaDoncila commented on docs/developers/napari_models.md:
```diff
@@ -0,0 +1,226 @@
+    """
+    Updates display for all sliders.
+
+    The event parameter is there just to allow easy connection to signals,
```
This sentence implies there should be an `event` in the method signature but we only have self?

lucyleeow replied:
Whoops left over from a re-write (I'm not actually completely clear on this, is it okay that we only have `self` here?)

DragaDoncila commented on docs/developers/napari_models.md:
```diff
@@ -0,0 +1,226 @@
+    The event parameter is there just to allow easy connection to signals,
+    without using `lambda event:`
+    """
+    # the code updating display code is quite complex and not relevant for this
```
```suggestion
    # the code updating the display code is not relevant for this
```

DragaDoncila commented on docs/developers/napari_models.md:
```diff
@@ -0,0 +1,226 @@
+    without using `lambda event:`
+    """
+    # the code updating display code is quite complex and not relevant for this
+    # example thus has ommited.
```
```suggestion
    # example thus has been omitted.
```

DragaDoncila commented on docs/developers/napari_models.md:
```diff
@@ -0,0 +1,226 @@
+    print(f"Update number of dimensions displayed to {nsteps}")
+
+# register our callback with the model
+dims.events.wind.connect(_update_display)
```
```suggestion
dims.events.ndisplay.connect(_update_display)
```

DragaDoncila commented on docs/developers/napari_models.md:
```diff
@@ -0,0 +1,226 @@
+"standardize" this change/emit pattern. The `EventedModel` provides the
+following features:
+
+* type validation and coersion on class instantiation and attribute assignment
```
```suggestion
* type validation and coercion on class instantiation and attribute assignment
```

DragaDoncila commented on docs/developers/napari_models.md:
```diff
@@ -0,0 +1,226 @@
+    """Vipsy camera for both 2D and 3D rendering.
+    """
+
+    def __init__(self, dims):
```
```suggestion
    def __init__(self, dims: Dims):
```

DragaDoncila reviewed (commented):
I really love this @lucyleeow ❤️ . I think you should definitely open this to napari - we can always add more detail and diagrams later. Have made mainly small comments and grammar/spelling corrections.

The only thing I was thinking is it might be nice to have a sentence or two at the very beginning about what this document will cover and who it's for/assumed knowledge e.g. "This document explains the links between the three main components of napari: python models, qt classes and vispy classes, with code examples. This document assumes you're familiar with blah blah."

lucyleeow committed:
address review

lucyleeow committed:
formatting

lucyleeow committed:
mv to guides and add toc tree

lucyleeow committed:
add section on headless

DragaDoncila committed:
Initialize plugins before validating widget in `napari -w {plugin}` (#4948)

* Initialize plugins before validating widget

* [pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

Co-authored-by: pre-commit-ci[bot] <66853113+pre-commit-ci[bot]@users.noreply.github.com>

lucyleeow committed:
remove add volume (#4939)

Since #493 add_volume no longer exists. This removes references to this function in the documentation.

Note I have also changed the name of files accordingly (e.g., add_volume to add_3D_image) and used 3D image instead of volume - though happy to change this back if not appropriate.

melissawm committed:
Remove repeated items in gallery ToC. (#4936)

Fixes napari/napari.github.io#361

As described in napari/napari.github.io#361, the items in the gallery appear multiple times in the sidebar ToC. I am not really sure when this behavior started, but this PR is meant to solve this problem.

melissawm committed:
Maintenance: Add title to example and remove unused redirect configuration (#4921)

The "Multiple viewer widget" example was added with no title, which was generating several sphinx warnings. In addition. this PR also removes the redirect configuration option from conf.py seeing as it is not working (and may cause trouble for local docs builds).

kandarpksk committed:
Correct minor typos on 'Contributing resources' page (#4927)

Benckmarks → Benchmarks
Core developer guide → Core Developer guide (consistent capitalization)

chili-chiu committed:
add SciPy 2022 materials (#4934)

jaimergp committed:
Use `napari/packaging` for Conda CI (#4923)

Following NAP-2, we are migrating the CI for packaging over to napari/packaging. However, due to the required cross-repository integrations, we will still run it from napari/napari thanks to the workflow_call triggers.

This allows to reuse the remote workflow with a napari/napari token, essentially. That way artifact publishing is still kept in this repo.

Carreau committed:
MAINT: temporary mypy disabling as it's failing everywhere. (#4950)

Merging this should be accompanied with opening an issue reminding us to
reenable

Carreau committed:
Merge branch 'main' into doc_arch

lucyleeow committed:
link to headless docs

lucyleeow committed:
review changes

lucyleeow committed:
review

lucyleeow commented:
closing as this is now outdated
